File: sdks/js/browser-sdk/src/utils/streamFailure.ts
const streamFailureMarker = "\n[XMTP_STREAM_FAILURE_V1]";
const maxCursor = (1n << 64n) - 1n;

export type StreamFailureCause = {
  kind:
    | "target_pending"
    | "receipt_pending"
    | "processing_pending"
    | "blocked"
    | "storage"
    | "receiver"
    | "invalid_topic";
  code: string | null;
  message: string;
  retryable: boolean;
};

export type UnfinishedStreamTopic = {
  topic: Uint8Array;
  /** Null means that target capture did not complete. */
  target: bigint | null;
  received: bigint;
  processed: bigint;
  unresolvedWelcomes: bigint[];
  inactive: boolean;
  cause: StreamFailureCause | null;
};

export type StreamBarrierFailure = {
  reason: "blocked" | "deadline" | "cancelled";
  unfinished: UnfinishedStreamTopic[];
};

export type StreamFailureDetails = {
  kind: "barrier" | "published_but_unconfirmed" | "catch_up";
  code: string;
  message: string;
  retryable: boolean;
  intentId: number | null;
  publishedIntentIds: number[];
  summary: {
    messages: bigint;
    conversations: bigint;
    failed: bigint;
    completed: boolean;
  } | null;
  barriers: StreamBarrierFailure[];
};

/**
 * Reads structured progress from a stream, sync, or publication error.
 * All sequence values are exact bigints. Invalid or absent details return undefined.
 * A published-but-unconfirmed error does not mean that publication failed.
 */
export const getStreamFailureDetails = (
  error: unknown,
): StreamFailureDetails | undefined => {
  try {
    const message =
      typeof error === "string" ? error : readObject(error).message;
    if (typeof message !== "string") return undefined;
    const marker = message.lastIndexOf(streamFailureMarker);
    if (marker === -1) return undefined;
    const details = readObject(
      JSON.parse(message.slice(marker + streamFailureMarker.length)),
    );
    return {
      kind: readChoice(details.kind, [
        "barrier",
        "published_but_unconfirmed",
        "catch_up",
      ]),
      code: readString(details.code),
      message: readString(details.message),
      retryable: readBoolean(details.retryable),
      intentId: readNullable(details.intentId, readIntentId),
      publishedIntentIds: readArray(details.publishedIntentIds, readIntentId),
      summary: readNullable(details.summary, (value) => {
        const summary = readObject(value);
        return {
          messages: readCursor(summary.messages),
          conversations: readCursor(summary.conversations),
          failed: readCursor(summary.failed),
          completed: readBoolean(summary.completed),
        };
      }),
      barriers: readArray(details.barriers, readBarrier),
    };
  } catch {
    return undefined;
  }
};

const readBarrier = (value: unknown): StreamBarrierFailure => {
  const barrier = readObject(value);
  return {
    reason: readChoice(barrier.reason, ["blocked", "deadline", "cancelled"]),
    unfinished: readArray(barrier.unfinished, (value) => {
      const topic = readObject(value);
      return {
        topic: readTopic(topic.topic),
        target: readNullable(topic.target, readCursor),
        received: readCursor(topic.received),
        processed: readCursor(topic.processed),
        unresolvedWelcomes: readArray(topic.unresolvedWelcomes, readCursor),
        inactive: readBoolean(topic.inactive),
        cause: readNullable(topic.cause, (value) => {
          const cause = readObject(value);
          return {
            kind: readChoice(cause.kind, [
              "target_pending",
              "receipt_pending",
              "processing_pending",
              "blocked",
              "storage",
              "receiver",
              "invalid_topic",
            ]),
            code: readNullable(cause.code, readString),
            message: readString(cause.message),
            retryable: readBoolean(cause.retryable),
          };
        }),
      };
    }),
  };
};

const readObject = (value: unknown): Record<string, unknown> => {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new Error("Invalid stream failure object");
  }
  return value as Record<string, unknown>;
};

const readString = (value: unknown): string => {
  if (typeof value !== "string") throw new Error("Invalid stream failure text");
  return value;
};

const readBoolean = (value: unknown): boolean => {
  if (typeof value !== "boolean")
    throw new Error("Invalid stream failure flag");
  return value;
};

const readChoice = <T extends string>(
  value: unknown,
  choices: readonly T[],
): T => {
  const choice = choices.find((choice) => choice === value);
  if (choice === undefined) throw new Error("Invalid stream failure kind");
  return choice;
};

const readNullable = <T>(
  value: unknown,
  read: (value: unknown) => T,
): T | null => (value === null ? null : read(value));

const readArray = <T>(value: unknown, read: (value: unknown) => T): T[] => {
  if (!Array.isArray(value)) throw new Error("Invalid stream failure list");
  return value.map(read);
};

const readCursor = (value: unknown): bigint => {
  const cursor = readString(value);
  if (!/^(0|[1-9][0-9]*)$/.test(cursor)) {
    throw new Error("Invalid stream failure sequence");
  }
  const number = BigInt(cursor);
  if (number > maxCursor)
    throw new Error("Stream failure sequence is too large");
  return number;
};

const readIntentId = (value: unknown): number => {
  if (
    typeof value !== "number" ||
    !Number.isInteger(value) ||
    value < -2147483648 ||
    value > 2147483647
  ) {
    throw new Error("Invalid stream failure intent ID");
  }
  return value;
};

const readTopic = (value: unknown): Uint8Array => {
  const hex = readString(value);
  if (!/^(?:[0-9a-fA-F]{2})*$/.test(hex)) {
    throw new Error("Invalid stream failure topic");
  }
  const bytes = new Uint8Array(hex.length / 2);
  for (let index = 0; index < bytes.length; index++) {
    bytes[index] = Number.parseInt(hex.slice(index * 2, index * 2 + 2), 16);
  }
  return bytes;
};

File: crates/xmtp_mls/src/subscriptions/stream_failure.rs
//! Structured processing failures shared by string-based binding errors.

use std::error::Error;

use serde::{Deserialize, Serialize};
use xmtp_common::{ErrorCode, RetryableError};

#[cfg(not(target_arch = "wasm32"))]
use super::catch_up::CatchUpError;
use super::{
    SubscribeError,
    barrier::{BarrierCause, BarrierError, BarrierFailure, BarrierTopic},
    incoming::IncomingError,
};
use crate::{
    client::ClientError,
    groups::{GroupError, mls_sync::GroupMessageProcessingError, summary::SyncSummary},
};

/// The suffix version is independent of the normal error message and code.
pub const STREAM_FAILURE_MARKER: &str = "\n[XMTP_STREAM_FAILURE_V1]";

/// The operation whose fixed processing obligations remain unfinished.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum StreamFailureKind {
    Barrier,
    PublishedButUnconfirmed,
    CatchUp,
}

/// Why the barrier stopped waiting. Durable pending work is retained.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum StreamBarrierReason {
    Blocked,
    Deadline,
    Cancelled,
}

/// Stable categories for one unfinished topic's cause.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum StreamBarrierCauseKind {
    TargetPending,
    ReceiptPending,
    ProcessingPending,
    Blocked,
    Storage,
    Receiver,
    InvalidTopic,
}

/// Safe cause metadata. Messages do not contain raw nested error data.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StreamBarrierCause {
    pub kind: StreamBarrierCauseKind,
    /// Stable source code when the cause has one.
    pub code: Option<String>,
    /// Fixed category text for display, not for control flow.
    pub message: String,
    pub retryable: bool,
}

/// All 64-bit values are decimal strings. Topic bytes use hexadecimal.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StreamBarrierTopic {
    /// Complete topic bytes, encoded as hexadecimal.
    pub topic: String,
    /// Fixed target H. None means capture failed; "0" is a captured empty target.
    pub target: Option<String>,
    /// Durable receipt cursor F. This is not application delivery progress.
    pub received: String,
    /// Resolved processing cursor P, independent of local delivery acknowledgement D.
    pub processed: String,
    /// Exact unresolved Welcome sequence IDs at or below H.
    pub unresolved_welcomes: Vec<String>,
    pub inactive: bool,
    pub cause: Option<StreamBarrierCause>,
}

/// One failed barrier and every topic obligation that it did not complete.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StreamBarrierFailure {
    pub reason: StreamBarrierReason,
    /// Keep all entries, including sibling failures from the same sync summary.
    pub unfinished: Vec<StreamBarrierTopic>,
}

/// Partial committed catch-up counts, encoded as exact decimal u64 strings.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StreamFailureSummary {
    pub messages: String,
    pub conversations: String,
    pub failed: String,
    pub completed: bool,
}

/// Versioned error details shared by all string-based binding error carriers.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StreamFailureDetails {
    pub kind: StreamFailureKind,
    pub code: String,
    pub message: String,
    pub retryable: bool,
    /// The sole published intent, if there is exactly one.
    pub intent_id: Option<i32>,
    /// All published intent IDs from nested errors, without duplicates.
    pub published_intent_ids: Vec<i32>,
    /// Committed catch-up progress when the error came from a catch-up run.
    pub summary: Option<StreamFailureSummary>,
    /// All failed barriers, not only the first source in an error chain.
    pub barriers: Vec<StreamBarrierFailure>,
}

impl From<&BarrierCause> for StreamBarrierCause {
    fn from(cause: &BarrierCause) -> Self {
        use StreamBarrierCauseKind as Kind;
        let (kind, code, message, retryable) = match cause {
            BarrierCause::TargetPending => {
                (Kind::TargetPending, None, "Target capture is pending", true)
            }
            BarrierCause::ReceiptPending => {
                (Kind::ReceiptPending, None, "Receipt is pending", true)
            }
            BarrierCause::ProcessingPending => {
                (Kind::ProcessingPending, None, "Processing is pending", true)
            }
            BarrierCause::Blocked(code) => (
                Kind::Blocked,
                Some(code.clone()),
                "Processing is blocked",
                false,
            ),
            BarrierCause::Storage(error) => (
                Kind::Storage,
                Some(error.error_code().to_string()),
                "Storage failed",
                error.is_retryable(),
            ),
            BarrierCause::Receiver(error) => (
                Kind::Receiver,
                Some(error.code().to_owned()),
                "The receiver failed",
                error.is_retryable(),
            ),
            BarrierCause::InvalidTopic => (Kind::InvalidTopic, None, "The topic is invalid", false),
        };
        Self {
            kind,
            code,
            message: message.to_owned(),
            retryable,
        }
    }
}

impl From<&BarrierTopic> for StreamBarrierTopic {
    fn from(topic: &BarrierTopic) -> Self {
        Self {
            topic: hex::encode(topic.topic.cloned_vec()),
            target: topic.target.map(|cursor| cursor.0.to_string()),
            received: topic.received.0.to_string(),
            processed: topic.processed.0.to_string(),
            unresolved_welcomes: topic
                .unresolved_welcomes
                .iter()
                .map(|cursor| cursor.0.to_string())
                .collect(),
            inactive: topic.inactive,
            cause: topic.cause.as_ref().map(Into::into),
        }
    }
}

impl From<&BarrierError> for StreamBarrierFailure {
    fn from(error: &BarrierError) -> Self {
        let BarrierError::Incomplete { reason, unfinished } = error;
        Self {
            reason: match reason {
                BarrierFailure::Blocked => StreamBarrierReason::Blocked,
                BarrierFailure::Deadline => StreamBarrierReason::Deadline,
                BarrierFailure::Cancelled => StreamBarrierReason::Cancelled,
            },
            unfinished: unfinished.iter().map(Into::into).collect(),
        }
    }
}

impl StreamFailureDetails {
    fn barrier(error: &BarrierError) -> Self {
        Self {
            kind: StreamFailureKind::Barrier,
            code: error.error_code().to_string(),
            message: "Processing barriers did not complete".into(),
            retryable: error.is_retryable(),
            intent_id: None,
            published_intent_ids: Vec::new(),
            summary: None,
            barriers: vec![error.into()],
        }
    }
}

/// Read known wrapper fields explicitly. Transparent errors can skip them in `source()`.
/// Visit every summary branch so a first error cannot hide sibling obligations.
pub fn stream_failure_details(error: &(dyn Error + 'static)) -> Option<StreamFailureDetails> {
    let mut pending = vec![(error, 0usize)];
    let mut failures = Vec::new();
    while let Some((error, depth)) = pending.pop() {
        if depth >= 64 {
            continue;
        }
        let mut push = |error| pending.push((error, depth + 1));
        #[cfg(not(target_arch = "wasm32"))]
        if let Some(error) = error.downcast_ref::<CatchUpError>() {
            match error {
                CatchUpError::Group(group) => push(group),
                CatchUpError::Incomplete { summary, causes } => {
                    failures.push(StreamFailureDetails {
                        kind: StreamFailureKind::CatchUp,
                        code: error.error_code().to_string(),
                        message: "Catch-up did not complete".into(),
                        retryable: error.is_retryable(),
                        intent_id: None,
                        published_intent_ids: Vec::new(),
                        summary: Some(StreamFailureSummary {
                            messages: summary.messages.to_string(),
                            conversations: summary.conversations.to_string(),
                            failed: summary.failed.to_string(),
                            completed: summary.completed,
                        }),
                        barriers: causes.iter().map(Into::into).collect(),
                    })
                }
            }
            continue;
        }
        if let Some(error) = error.downcast_ref::<GroupError>() {
            match error {
                GroupError::StreamBarrier(error) => {
                    failures.push(StreamFailureDetails::barrier(error))
                }
                GroupError::PublishedButUnconfirmed { intent_id, cause } => {
                    failures.push(StreamFailureDetails {
                        kind: StreamFailureKind::PublishedButUnconfirmed,
                        code: error.error_code().to_string(),
                        message: "The published intent is not confirmed".into(),
                        retryable: error.is_retryable(),
                        intent_id: Some(*intent_id),
                        published_intent_ids: vec![*intent_id],
                        summary: None,
                        barriers: cause.iter().map(|cause| cause.as_ref().into()).collect(),
                    })
                }
                GroupError::Sync(summary) | GroupError::SyncFailedToWait(summary) => {
                    push(summary.as_ref())
                }
                GroupError::Client(error) => push(error),
                _ => {
                    if let Some(source) = error.source() {
                        push(source);
                    }
                }
            }
            continue;
        }
        if let Some(error) = error.downcast_ref::<BarrierError>() {
            failures.push(StreamFailureDetails::barrier(error));
            continue;
        }
        if let Some(summary) = error.downcast_ref::<SyncSummary>() {
            for (_, error) in summary.process.errored.iter().rev() {
                push(error);
            }
            for error in summary.post_commit_errors.iter().rev() {
                push(error);
            }
            for error in summary.publish_errors.iter().rev() {
                push(error);
            }
            if let Some(error) = summary.other.as_deref() {
                push(error);
            }
            continue;
        }
        if let Some(ClientError::Group(error)) = error.downcast_ref::<ClientError>() {
            push(error.as_ref());
        } else if let Some(SubscribeError::Group(error)) = error.downcast_ref::<SubscribeError>() {
            push(error.as_ref());
        } else if let Some(IncomingError::Group(error)) = error.downcast_ref::<IncomingError>() {
            push(error);
        } else if let Some(error) = error.downcast_ref::<GroupMessageProcessingError>() {
            match error {
                GroupMessageProcessingError::PreparedAttempt(error) => push(error.as_ref()),
                GroupMessageProcessingError::Client(error) => push(error),
                _ => {
                    if let Some(source) = error.source() {
                        push(source);
                    }
                }
            }
        } else if let Some(source) = error.source() {
            push(source);
        }
    }
    let mut failures = failures.into_iter();
    let mut result = failures.next()?;
    for failure in failures {
        result.retryable |= failure.retryable;
        result.barriers.extend(failure.barriers);
        result
            .published_intent_ids
            .extend(failure.published_intent_ids);
    }
    result.published_intent_ids.sort_unstable();
    result.published_intent_ids.dedup();
    result.intent_id = match result.published_intent_ids.as_slice() {
        [id] => Some(*id),
        _ => None,
    };
    Some(result)
}

/// Append this suffix after the existing error code and message.
pub fn encode_stream_failure(error: &(dyn Error + 'static)) -> Option<String> {
    let details = stream_failure_details(error)?;
    let json = serde_json::to_string(&details).ok()?;
    Some(format!("{STREAM_FAILURE_MARKER}{json}"))
}

/// Decode the versioned suffix. Ordinary errors have no structured details.
pub fn decode_stream_failure(message: &str) -> Option<StreamFailureDetails> {
    let (_, json) = message.rsplit_once(STREAM_FAILURE_MARKER)?;
    serde_json::from_str(json).ok()
}

#[cfg(test)]
mod tests;
